### DISCUSSION TITLE
OpenSSL 3.2.3 aes-128-gcm

### DISCUSSION THREAD

**Discussion by sku-schleupen** [openssl/openssl#26148] [Q&A]
Part of my work is in the German EDIFACT environment and starting on the 01.04.25, Mails sent for EDIFACT need to be encrypted using the algorithm AES-128-GCM.
Some marketpartners are already encrypting using the new algorithm, and so far we are not able to decrypt these with our software.
I have tried using openssl directly to decrypt the mails, but I simply get "Error decrypting CMS structure".
It is hard to say if the issue lies with the encryptor or the decryptor (apart from the fact that the algorithm shouldn't be used yet).
Are there any known issues with OpenSSL and AES-128-GCM?

`authEncryptedContentInfo:
      contentType: pkcs7-data (1.2.840.113549.1.7.1)
      contentEncryptionAlgorithm:
        algorithm: aes-128-gcm (2.16.840.1.101.3.4.1.6)
        parameter: SEQUENCE:
    0:d=0  hl=2 l=  14 cons: SEQUENCE
    2:d=1  hl=2 l=  12 prim:  OCTET STRING      [HEX DUMP]:D4E87D1CB3CADCACAED17AF7`

**Comment by mattcaswell** [comment]
I am successfully able to use aes-128-gcm using the `openssl cms` app:

````
$ cat test.txt
The quick brown fox jumps over the lazy dog
$ openssl cms -encrypt -in test.txt -aes-128-gcm -out test.enc -outform pem cert.pem
$ openssl cms -decrypt -in test.enc -inform pem -out test-out.txt -inkey key.pem 
$ cat test-out.txt
The quick brown fox jumps over the lazy dog
$ openssl asn1parse -in test.enc
    0:d=0  hl=4 l= 445 cons: SEQUENCE          
    4:d=1  hl=2 l=  11 prim: OBJECT            :id-smime-ct-authEnvelopedData
   17:d=1  hl=4 l= 428 cons: cont [ 0 ]        
   21:d=2  hl=4 l= 424 cons: SEQUENCE          
   25:d=3  hl=2 l=   1 prim: INTEGER           :00
   28:d=3  hl=4 l= 307 cons: SET               
   32:d=4  hl=4 l= 303 cons: SEQUENCE          
   36:d=5  hl=2 l=   1 prim: INTEGER           :00
   39:d=5  hl=2 l=  23 cons: SEQUENCE          
   41:d=6  hl=2 l=  18 cons: SEQUENCE          
   43:d=7  hl=2 l=  16 cons: SET               
   45:d=8  hl=2 l=  14 cons: SEQUENCE          
   47:d=9  hl=2 l=   3 prim: OBJECT            :commonName
   52:d=9  hl=2 l=   7 prim: UTF8STRING        :Root CA
   61:d=6  hl=2 l=   1 prim: INTEGER           :02
   64:d=5  hl=2 l=  13 cons: SEQUENCE          
   66:d=6  hl=2 l=   9 prim: OBJECT            :rsaEncryption
   77:d=6  hl=2 l=   0 prim: NULL              
   79:d=5  hl=4 l= 256 prim: OCTET STRING      [HEX DUMP]:8565F710D9B8058AB373EC3DA2B5EB4A2F9EA7BC41EF56BCAF427AFFF094BEDC76BD4248DE9A1392CFEF70ABA1EFCC5802345BBB2B2C489ED6654C0DA6CBE529466ECD38819A19842EC0396A76097ACF3620B23E6A5A72368CA9B0487B1FB49AD23AA5BEFF50075CD8B18FAFA193E14F69B1C1EC956213514767A8A14761D9266E082688B3D89E20B8A498433801626884B7A92F14A586E185E4F1D74B4D9B3AEFA24F60F6BDBA84F3542B90ED2ACD6B65D6B467353A2ADA56FF65C1C478829EDB037052763339D23CC8FFF3979666820F9D57081A287C1372446A0692C9912135DA67A5636FF5257F6C110861B23D544F2CE5851B9E2CEA3EC15CCF240B12BE
  339:d=3  hl=2 l=  90 cons: SEQUENCE          
  341:d=4  hl=2 l=   9 prim: OBJECT            :pkcs7-data
  352:d=4  hl=2 l=  30 cons: SEQUENCE          
  354:d=5  hl=2 l=   9 prim: OBJECT            :aes-128-gcm
  365:d=5  hl=2 l=  17 cons: SEQUENCE          
  367:d=6  hl=2 l=  12 prim: OCTET STRING      [HEX DUMP]:957C841BF1E105713F488D03
  381:d=6  hl=2 l=   1 prim: INTEGER           :10
  384:d=4  hl=2 l=  45 prim: cont [ 0 ]        
  431:d=3  hl=2 l=  16 prim: OCTET STRING      [HEX DUMP]:C4B5407975299B97F01F529C5A89C946
$ openssl version
OpenSSL 3.0.13 30 Jan 2024 (Library: OpenSSL 3.0.13 30 Jan 2024)
````

What version of OpenSSL are you using? How are you attempting to decrypt, e.g. via the openssl cms app, or programmatically?

**Comment by sku-schleupen** [comment]
Decrypting aes-128-gcm using openssl which was encrypted using openssl is not an issue, we have found out.

The error only occurs when another API is used, such as bouncycastle.

There seems to be some incompatibility.

OpenSSL 3.2.3 3 Sep 2024 (Library: OpenSSL 3.2.3 3 Sep 2024)

I decrypt using cli:
`openssl cms -decrypt -recip edi-netz.pem -in message.mail -out decrypt.mail`

**Reply by mattcaswell** [reply]
Are you able to provide some test cert / key / message that you can share here?

**Comment by erdbeerschnitzel** [comment]
This looks like the topic discussed here: https://github.com/openssl/openssl/issues/21414
